File: app/src/main/java/com/UIN/Tool/utils/UpdateDownloader.java
package com.UIN.Tool.utils;

import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.Environment;
import android.os.Handler;
import android.os.Looper;
import android.widget.Toast;

import androidx.core.app.NotificationCompat;
import androidx.core.content.FileProvider;

import com.UIN.Tool.MainActivity;
import com.UIN.Tool.R;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.security.cert.X509Certificate;

import javax.net.ssl.HostnameVerifier;
import javax.net.ssl.HttpsURLConnection;
import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLSession;
import javax.net.ssl.TrustManager;
import javax.net.ssl.X509TrustManager;

public class UpdateDownloader {
    
    private static final String TAG = "UpdateDownloader";
    private static final String CHANNEL_ID = "update_download_channel";
    private static final String CHANNEL_NAME = "更新下载";
    private static final int NOTIFICATION_ID = 1001;
    
    private Context context;
    private NotificationManager notificationManager;
    private NotificationCompat.Builder notificationBuilder;
    private Handler mainHandler;
    private OnDownloadListener listener;
    private boolean isDownloading = false;
    
    public interface OnDownloadListener {
        void onStart();
        void onProgress(int progress, long downloaded, long total);
        void onSuccess(File file);
        void onFailed(String error);
    }
    
    public UpdateDownloader(Context context) {
        this.context = context.getApplicationContext();
        this.mainHandler = new Handler(Looper.getMainLooper());
        this.notificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        createNotificationChannel();
        disableSSLCertificateChecking();
        LogUtils.i(TAG, "UpdateDownloader 初始化完成");
    }
    
    private void disableSSLCertificateChecking() {
        try {
            TrustManager[] trustAllCerts = new TrustManager[]{
                new X509TrustManager() {
                    @Override
                    public void checkClientTrusted(X509Certificate[] chain, String authType) {}
                    @Override
                    public void checkServerTrusted(X509Certificate[] chain, String authType) {}
                    @Override
                    public X509Certificate[] getAcceptedIssuers() {
                        return new X509Certificate[]{};
                    }
                }
            };
            
            SSLContext sslContext = SSLContext.getInstance("TLS");
            sslContext.init(null, trustAllCerts, new java.security.SecureRandom());
            HttpsURLConnection.setDefaultSSLSocketFactory(sslContext.getSocketFactory());
            
            HostnameVerifier allHostsValid = (hostname, session) -> true;
            HttpsURLConnection.setDefaultHostnameVerifier(allHostsValid);
            
            LogUtils.i(TAG, "SSL 证书检查已禁用");
        } catch (Exception e) {
            LogUtils.e(TAG, "禁用 SSL 检查失败", e);
        }
    }
    
    private void createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel channel = new NotificationChannel(
                CHANNEL_ID,
                CHANNEL_NAME,
                NotificationManager.IMPORTANCE_LOW
            );
            channel.setDescription("应用更新下载进度");
            notificationManager.createNotificationChannel(channel);
        }
    }
    
    public void setOnDownloadListener(OnDownloadListener listener) {
        this.listener = listener;
    }
    
    public boolean isDownloading() {
        return isDownloading;
    }
    
    public void cancelDownload() {
        isDownloading = false;
        if (notificationBuilder != null) {
            notificationBuilder.setProgress(0, 0, false);
            notificationBuilder.setContentText("下载已取消");
            notificationManager.notify(NOTIFICATION_ID, notificationBuilder.build());
        }
        mainHandler.postDelayed(() -> notificationManager.cancel(NOTIFICATION_ID), 2000);
    }
    
    public void startDownload(String downloadUrl, String versionName) {
        if (isDownloading) {
            showToast("正在下载中，请稍后...");
            return;
        }
        
        isDownloading = true;
        
        if (listener != null) {
            listener.onStart();
        }
        
        createNotification(versionName);
        
        new Thread(new DownloadRunnable(downloadUrl, versionName)).start();
    }
    
    private void createNotification(String versionName) {
        Intent intent = new Intent(context, MainActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
        PendingIntent pendingIntent = PendingIntent.getActivity(
            context, 0, intent,
            PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE
        );
        
        notificationBuilder = new NotificationCompat.Builder(context, CHANNEL_ID)
            .setContentTitle("正在下载 UIN Tool " + versionName)
            .setContentText("下载中... 0%")
            .setSmallIcon(android.R.drawable.stat_sys_download)
            .setContentIntent(pendingIntent)
            .setOngoing(true)
            .setPriority(NotificationCompat.PRIORITY_LOW);
        
        notificationManager.notify(NOTIFICATION_ID, notificationBuilder.build());
    }
    
    private void updateNotification(int progress, long downloaded, long total) {
        if (notificationBuilder == null) return;
        
        String downloadedStr = formatSize(downloaded);
        String totalStr = formatSize(total);
        
        notificationBuilder
            .setProgress(100, progress, false)
            .setContentText("下载中... " + progress + "% (" + downloadedStr + "/" + totalStr + ")");
        
        notificationManager.notify(NOTIFICATION_ID, notificationBuilder.build());
    }
    
    /**
     * 安装 APK 文件
     */
    public void installApk(File apkFile) {
        if (apkFile == null || !apkFile.exists()) {
            LogUtils.e(TAG, "APK 文件不存在");
            showToast("安装包不存在");
            return;
        }
        
        try {
            Intent intent = new Intent(Intent.ACTION_VIEW);
            Uri apkUri;
            
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                apkUri = FileProvider.getUriForFile(context, context.getPackageName() + ".fileprovider", apkFile);
                intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
            } else {
                apkUri = Uri.fromFile(apkFile);
            }
            
            intent.setDataAndType(apkUri, "application/vnd.android.package-archive");
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            context.startActivity(intent);
            
            LogUtils.success(TAG, "启动安装: " + apkFile.getAbsolutePath());
        } catch (Exception e) {
            LogUtils.e(TAG, "启动安装失败", e);
            showToast("无法启动安装: " + e.getMessage());
        }
    }
    
    private String formatSize(long size) {
        if (size <= 0) return "0 B";
        if (size < 1024) return size + " B";
        if (size < 1024 * 1024) return String.format("%.2f KB", size / 1024.0);
        return String.format("%.2f MB", size / (1024.0 * 1024.0));
    }
    
    private void showToast(final String message) {
        mainHandler.post(() -> Toast.makeText(context, message, Toast.LENGTH_SHORT).show());
    }
    
    private class DownloadRunnable implements Runnable {
        private String downloadUrl;
        private String versionName;
        
        DownloadRunnable(String downloadUrl, String versionName) {
            this.downloadUrl = downloadUrl;
            this.versionName = versionName;
        }
        
        @Override
        public void run() {
            HttpURLConnection connection = null;
            InputStream inputStream = null;
            FileOutputStream outputStream = null;
            
            try {
                File downloadDir = new File(Environment.getExternalStorageDirectory(), "UIN_Tool/downloads");
                if (!downloadDir.exists()) {
                    downloadDir.mkdirs();
                }
                
                String cleanVersion = versionName;
                if (cleanVersion.startsWith("v") || cleanVersion.startsWith("V")) {
                    cleanVersion = cleanVersion.substring(1);
                }
                File apkFile = new File(downloadDir, "UIN_Tool_v" + cleanVersion + ".apk");
                
                if (apkFile.exists()) {
                    apkFile.delete();
                }
                
                LogUtils.i(TAG, "开始下载: " + downloadUrl);
                LogUtils.i(TAG, "保存路径: " + apkFile.getAbsolutePath());
                
                URL url = new URL(downloadUrl);
                connection = (HttpURLConnection) url.openConnection();
                connection.setConnectTimeout(30000);
                connection.setReadTimeout(30000);
                connection.setRequestMethod("GET");
                connection.setRequestProperty("User-Agent", "UIN-Tool-Android");
                
                int responseCode = connection.getResponseCode();
                if (responseCode != HttpURLConnection.HTTP_OK) {
                    throw new Exception("服务器响应错误: " + responseCode);
                }
                
                long totalSize = connection.getContentLength();
                LogUtils.param(TAG, "文件大小", formatSize(totalSize));
                
                inputStream = connection.getInputStream();
                outputStream = new FileOutputStream(apkFile);
                
                byte[] buffer = new byte[8192];
                long downloadedSize = 0;
                int bytesRead;
                int lastProgress = -1;
                
                while ((bytesRead = inputStream.read(buffer)) != -1 && isDownloading) {
                    outputStream.write(buffer, 0, bytesRead);
                    downloadedSize += bytesRead;
                    
                    if (totalSize > 0) {
                        final int progress = (int) (downloadedSize * 100 / totalSize);
                        
                        if (progress != lastProgress) {
                            lastProgress = progress;
                            
                            final long finalDownloaded = downloadedSize;
                            final long finalTotal = totalSize;
                            
                            mainHandler.post(() -> {
                                if (listener != null) {
                                    listener.onProgress(progress, finalDownloaded, finalTotal);
                                }
                                updateNotification(progress, finalDownloaded, finalTotal);
                            });
                            
                            LogUtils.d(TAG, "下载进度: " + progress + "% (" + formatSize(downloadedSize) + "/" + formatSize(totalSize) + ")");
                        }
                    }
                }
                
                outputStream.flush();
                
                if (!isDownloading) {
                    apkFile.delete();
                    mainHandler.post(() -> {
                        if (listener != null) {
                            listener.onFailed("下载已取消");
                        }
                    });
                    return;
                }
                
                LogUtils.success(TAG, "下载完成: " + apkFile.getAbsolutePath());
                
                mainHandler.post(() -> {
                    if (listener != null) {
                        listener.onSuccess(apkFile);
                    }
                    notificationManager.cancel(NOTIFICATION_ID);
                });
                
            } catch (final Exception e) {
                LogUtils.e(TAG, "下载失败", e);
                mainHandler.post(() -> {
                    if (listener != null) {
                        listener.onFailed(e.getMessage());
                    }
                    notificationManager.cancel(NOTIFICATION_ID);
                });
            } finally {
                isDownloading = false;
                try {
                    if (inputStream != null) inputStream.close();
                    if (outputStream != null) outputStream.close();
                    if (connection != null) connection.disconnect();
                } catch (Exception e) {
                    LogUtils.e(TAG, "关闭流失败", e);
                }
            }
        }
    }
}

File: app/src/main/java/com/UIN/Tool/core/update/UpdateDownloader.kt
package com.UIN.Tool.core.update

import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Environment
import androidx.core.app.NotificationCompat
import androidx.core.content.FileProvider
import com.UIN.Tool.MainActivity
import com.UIN.Tool.log.Logger
import com.UIN.Tool.utils.Constants
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.File
import java.io.FileOutputStream
import java.util.concurrent.TimeUnit

class UpdateDownloader(
    private val context: Context
) {

    companion object {
        private const val TAG = "UpdateDownloader"
        private const val CHANNEL_ID = "update_download_channel"
        private const val NOTIFICATION_ID = 1001
    }

    private var onDownloadListener: OnDownloadListener? = null
    private var isDownloading = false
    private var downloadThread: Thread? = null

    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(60, TimeUnit.SECONDS)
        .writeTimeout(60, TimeUnit.SECONDS)
        .build()

    private val notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

    interface OnDownloadListener {
        fun onStart()
        fun onProgress(progress: Int, downloaded: Long, total: Long)
        fun onSuccess(file: File)
        fun onFailed(error: String)
    }

    init {
        createNotificationChannel()
    }

    fun setOnDownloadListener(listener: OnDownloadListener) {
        this.onDownloadListener = listener
    }

    private fun createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(
                CHANNEL_ID,
                "更新下载",
                NotificationManager.IMPORTANCE_LOW
            ).apply {
                description = "应用更新下载进度"
                setShowBadge(false)
            }
            notificationManager.createNotificationChannel(channel)
        }
    }

    fun startDownload(downloadUrl: String, versionName: String) {
        if (isDownloading) {
            Logger.w(TAG, "下载正在进行中")
            return
        }

        isDownloading = true
        onDownloadListener?.onStart()

        // 创建下载目录
        val downloadDir = File(Constants.DOWNLOAD_DIR)
        if (!downloadDir.exists()) {
            downloadDir.mkdirs()
        }

        // 清理版本号前缀
        val cleanVersion = versionName.replace(Regex("^[vV]"), "")
        val fileName = "UIN_Tool_v${cleanVersion}.apk"
        val file = File(downloadDir, fileName)

        if (file.exists()) {
            file.delete()
        }

        Logger.i(TAG, "开始下载: $downloadUrl")
        Logger.param(TAG, "保存路径", file.absolutePath)

        // 显示通知
        showNotification(versionName)

        downloadThread = Thread {
            try {
                val request = Request.Builder()
                    .url(downloadUrl)
                    .header("User-Agent", "UIN-Tool-Android")
                    .build()

                val response = client.newCall(request).execute()
                if (!response.isSuccessful) {
                    throw Exception("下载失败: ${response.code}")
                }

                val body = response.body ?: throw Exception("响应体为空")
                val contentLength = body.contentLength()

                FileOutputStream(file).use { fos ->
                    body.byteStream().use { inputStream ->
                        val buffer = ByteArray(8192)
                        var bytesRead: Int
                        var downloaded = 0L
                        var lastProgress = -1

                        while (inputStream.read(buffer).also { bytesRead = it } != -1) {
                            // 检查是否被取消
                            if (!isDownloading) {
                                fos.close()
                                file.delete()
                                throw Exception("下载已取消")
                            }

                            fos.write(buffer, 0, bytesRead)
                            downloaded += bytesRead

                            val progress = if (contentLength > 0) {
                                (downloaded * 100 / contentLength).toInt()
                            } else 0

                            if (progress != lastProgress) {
                                lastProgress = progress
                                updateNotification(progress, downloaded, contentLength)
                                onDownloadListener?.onProgress(progress, downloaded, contentLength)
                                Logger.d(TAG, "下载进度: $progress%")
                            }
                        }
                    }
                }

                Logger.success(TAG, "下载完成: ${file.absolutePath}")

                notificationManager.cancel(NOTIFICATION_ID)
                isDownloading = false
                onDownloadListener?.onSuccess(file)

            } catch (e: Exception) {
                Logger.e(TAG, "下载失败", e)
                notificationManager.cancel(NOTIFICATION_ID)
                isDownloading = false
                onDownloadListener?.onFailed(e.message ?: "下载失败")
            }
        }

        downloadThread?.start()
    }

    fun cancelDownload() {
        isDownloading = false
        downloadThread?.interrupt()
        notificationManager.cancel(NOTIFICATION_ID)
        Logger.i(TAG, "下载已取消")
    }

    fun installApk(file: File) {
        if (!file.exists()) {
            Logger.e(TAG, "APK文件不存在: ${file.absolutePath}")
            return
        }

        try {
            val uri = FileProvider.getUriForFile(
                context,
                "${context.packageName}.fileprovider",
                file
            )

            val intent = Intent(Intent.ACTION_VIEW).apply {
                setDataAndType(uri, "application/vnd.android.package-archive")
                addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            }

            context.startActivity(intent)
            Logger.success(TAG, "启动安装: ${file.name}")
        } catch (e: Exception) {
            Logger.e(TAG, "启动安装失败", e)
        }
    }

    private fun showNotification(versionName: String) {
        val intent = Intent(context, MainActivity::class.java).apply {
            flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP
        }
        val pendingIntent = PendingIntent.getActivity(
            context,
            0,
            intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )

        val notification = NotificationCompat.Builder(context, CHANNEL_ID)
            .setContentTitle("正在下载 UIN Tool $versionName")
            .setContentText("下载中... 0%")
            .setSmallIcon(android.R.drawable.stat_sys_download)
            .setContentIntent(pendingIntent)
            .setOngoing(true)
            .setPriority(NotificationCompat.PRIORITY_LOW)
            .setProgress(100, 0, false)
            .build()

        notificationManager.notify(NOTIFICATION_ID, notification)
    }

    private fun updateNotification(progress: Int, downloaded: Long, total: Long) {
        val downloadedStr = formatSize(downloaded)
        val totalStr = formatSize(total)

        val notification = NotificationCompat.Builder(context, CHANNEL_ID)
            .setContentTitle("正在下载更新")
            .setContentText("下载中... $progress% ($downloadedStr/$totalStr)")
            .setProgress(100, progress, false)
            .setSmallIcon(android.R.drawable.stat_sys_download)
            .setOngoing(true)
            .setPriority(NotificationCompat.PRIORITY_LOW)
            .build()

        notificationManager.notify(NOTIFICATION_ID, notification)
    }

    private fun formatSize(size: Long): String {
        return when {
            size <= 0 -> "0 B"
            size < 1024 -> "$size B"
            size < 1024 * 1024 -> String.format("%.2f KB", size / 1024.0)
            else -> String.format("%.2f MB", size / (1024.0 * 1024.0))
        }
    }
}